core/drivers/crypto/caam/hal/imx_6_7/hal_clk.c

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018-2019 NXP
+ *
+ * Brief   CAAM Clock functions.
+ */
+#include <caam_hal_clk.h>
+#include <caam_io.h>
+#include <mm/core_memprot.h>
+
+#if defined(CFG_MX6) || defined(CFG_MX6UL)
+void caam_hal_clk_enable(bool enable)
+{
+	vaddr_t ccm_base = (vaddr_t)phys_to_virt(CCM_BASE, MEM_AREA_IO_SEC);
+	uint32_t reg = 0;
+	uint32_t mask = 0;
+
+	reg = io_caam_read32(ccm_base + CCM_CCGR0);
+
+	mask = BM_CCM_CCGR0_CAAM_WRAPPER_IPG | BM_CCM_CCGR0_CAAM_WRAPPER_ACLK |
+	       BM_CCM_CCGR0_CAAM_SECURE_MEM;
+
+	if (enable)
+		reg |= mask;
+	else
+		reg &= ~mask;
+
+	io_caam_write32(ccm_base + CCM_CCGR0, reg);
+
+	if (!soc_is_imx6ul()) {
+		/* EMI slow clk */
+		reg = io_caam_read32(ccm_base + CCM_CCGR6);
+		mask = BM_CCM_CCGR6_EMI_SLOW;
+
+		if (enable)
+			reg |= mask;
+		else
+			reg &= ~mask;
+
+		io_caam_write32(ccm_base + CCM_CCGR6, reg);
+	}
+}
+#endif /* CFG_MX6 || CFG_MX6UL */
+
+#if defined(CFG_MX7)
+void caam_hal_clk_enable(bool enable)
+{
+	vaddr_t ccm_base = (vaddr_t)phys_to_virt(CCM_BASE, MEM_AREA_IO_SEC);
+
+	if (enable) {
+		io_caam_write32(ccm_base + CCM_CCGRx_SET(CCM_CLOCK_DOMAIN_CAAM),
+				CCM_CCGRx_ALWAYS_ON(0));
+	} else {
+		io_caam_write32(ccm_base + CCM_CCGRx_CLR(CCM_CLOCK_DOMAIN_CAAM),
+				CCM_CCGRx_ALWAYS_ON(0));
+	}
+}
+#endif /* CFG_MX7 */
+
+#if defined(CFG_MX7ULP)
+void caam_hal_clk_enable(bool enable)
+{
+	vaddr_t pcc2_base = (vaddr_t)phys_to_virt(PCC2_BASE, MEM_AREA_IO_SEC);
+
+	if (enable)
+		io_caam_write32(pcc2_base + PCC_CAAM, PCC_ENABLE_CLOCK);
+	else
+		io_caam_write32(pcc2_base + PCC_CAAM, PCC_DISABLE_CLOCK);
```

Comment:
Looks like this pull request is missing the definitions of PCC_ENABLE/DISABLE_CLOCK, which is provided by the imx7ulp-crm_regs.h file in the imx fork.

Reply:
Thanks for highlighting it. This is a fix that will be submitted with another patch when this PR will be merged.

Reply:
Isn't it just easier and better to send a patch adding that header file first (just the header)? Otherwise users trying to enable it will face a build failure.

Reply:
This CAAM Driver is not build for this platform 7ULP. I'll see if the file can be added in a new comnit even if not necessary for this current PR

Reply:
> This CAAM Driver is not build for this platform 7ULP

Then why is the code even here inside `#if defined(CFG_MX7ULP)`/`#endif`?

Also we said on many occasions that we prefer to avoid #ifdefs in `.c` files if possible. It seems to me you could create one file per platform and use the pattern `src-$(CFG_XYZ) += xyz.c` here, no?

Reply:
is it so mandatory for this case? 
The code is already in a hal folder specific for a category of device. For a so small function that is not complex we didn't estimate to have a file per specific iMX define.

Reply:
Mandatory, no, but it's just more convenient. For instance it's easier to navigate with grep or Vim ctags or whatever -- I'm showing my age here, perhaps you are using a more modern IDE, but anyway ;-) Say you're looking for the iMX6 clock enable function:

```
$ git grep caam_hal_clk_enable
hal_clk.c:void caam_hal_clk_enable(bool enable)
hal_clk.c:void caam_hal_clk_enable(bool enable)
hal_clk.c:void caam_hal_clk_enable(bool enable)
```
Which one would you pick?

Compare with:
```
$ git grep caam_hal_clk_enable
hal_clk_mx6_mx6ul.c:void caam_hal_clk_enable(bool enable)
hal_clk_mx7.c:void caam_hal_clk_enable(bool enable)
hal_clk_mx7ulp.c:void caam_hal_clk_enable(bool enable)
```

On the other hand, what's the advantage of having a single file?

Reply:
done